src/backend/distributed/relay/relay_event_utility.c

```diff
@@ -624,32 +631,120 @@ AppendShardIdToName(char **name, uint64 shardId)
 	}
 
+	/*
+	 * Otherwise, we need to truncate the name further to accommodate
+	 * a sufficient hash value. The resulting name will avoid collision
+	 * with other hashed names such that for any given schema with
+	 * 90 distinct object names that are long enough to require hashing
+	 * (typically 57-63 characters), the chance of a collision existing is:
+	 *
+	 * If randomly generated UTF8 names:
+	 *     (1e-6) * (9.39323783788e-114) ~= (9.39e-120)
+	 * If random case-insensitive ASCII names (letter first, 37 useful characters):
+	 *     (1e-6) * (2.80380202421e-74) ~= (2.8e-80)
+	 * If names sharing only N distinct 45- to 47-character prefixes:
+	 *     (1e-6) * (1/N) = (1e-6/N)
+	 *     1e-7 for 10 distinct prefixes
+	 *     5e-8 for 20 distinct prefixes
+	 *
+	 * In practice, since shard IDs are globally unique, the risk of name collision
+	 * exists only amongst objects that pertain to a single distributed table
+	 * and are created for each shard: the table name and the names of any indexes
+	 * or index-backed constraints. Since there are typically less than five such
+	 * names, and almost never more than ten, the expected collision rate even in
+	 * the worst case (ten names share same 45- to 47-character prefix) is roughly
+	 * 1e-8: one in 100 million schemas will experience a name collision only if ALL
+	 * 100 million schemas present the worst-case scenario.
+	 */
+	else
+	{
+		longNameHash = hash_any((unsigned char *) (*name), nameLength);
+		multiByteClipLength = pg_mbcliplen(*name, nameLength, (NAMEDATALEN -
+															   shardIdAndSeparatorLength -
+															   10));
+		snprintf(extendedName, NAMEDATALEN, "%.*s%c%.8x%s",
+				 multiByteClipLength, (*name),
+				 SHARD_NAME_SEPARATOR, longNameHash,
+				 shardIdAndSeparator);
+	}
+	extendedNameLength = strlen(extendedName) + 1;
+	Assert(extendedNameLength <= NAMEDATALEN);
+
 	(*name) = (char *) repalloc((*name), extendedNameLength);
 	snprintf((*name), extendedNameLength, "%s", extendedName);
 }
 
 
 /*
- * AppendShardIdToStringInfo appends shardId to the given name, represented
- * by a StringInfo.
+ * shard_name() provides a PG function interface to AppendShardNameToId above.
  */
-void
-AppendShardIdToStringInfo(StringInfo name, uint64 shardId)
+Datum
+shard_name(PG_FUNCTION_ARGS)
 {
-	appendStringInfo(name, "%c" UINT64_FORMAT, SHARD_NAME_SEPARATOR, shardId);
+	Oid relationId = InvalidOid;
+	uint64 shardId = 0;
+	char *relationName = NULL;
+
+	/*
+	 * Have to check arguments for NULLness as it can't be declared STRICT
+	 * because of min/max arguments, which have to be NULLable for new shards.
+	 */
+	if (PG_ARGISNULL(0))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("object_name cannot be null")));
+	}
+	if (PG_ARGISNULL(1))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("shard_id cannot be null")));
+	}
+
+
+	relationId = PG_GETARG_OID(0);
+	shardId = PG_GETARG_INT64(1);
+
+	if (shardId <= 0)
```

Comment:
But the argument signature allows bigint; just want to reject bad arguments for a function that users' scripts can call...